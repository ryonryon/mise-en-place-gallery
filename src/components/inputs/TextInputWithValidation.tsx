import styled from "@emotion/styled";
import * as React from "react";

import BasicTextInput from "./BasicTextInput";
import InputType from "./input-type";
import { primary, secondary, danger } from "../../constants/colors";

interface Props extends React.InputHTMLAttributes<HTMLInputElement> {
  type: InputType;
  validations: [RegExp | Function, string][];
}

const TextWithValidationMessages = styled.div`
  & > input {
    border-color: ${secondary};
    color: ${secondary};

    &::placeholder {
      color: ${secondary};
    }
  }

  & > .errorStyle {
    margin-bottom: 4px;
    border-color: ${danger};
  }

  & > .successStyle {
    border-color: ${primary};
  }

  & > ul {
    width: 70%;
    margin: 0 auto;
    padding-left: 30px;
  }

  & > ul > li {
    list-style: none;
    text-align: left;
    font-size: 14px;
    color: ${danger};
  }
`;

const TextInputWithValidation = ({ validations, ...rest }: Props) => {
  const [value, setValue] = React.useState("");
  const [errorMessages, setErrorMessages] = React.useState([] as String[]);
  const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();
    const val = e.currentTarget.value;
    setValue(val);

    setErrorMessages(
      validations
        .map(([validation, errorMessage]) => {
          if (typeof validation === "function") {
            if (validation(val)) return errorMessage;
          } else {
            if (val.match(validation)) return errorMessage;
          }
        })
        .filter(item => item !== undefined)
    );
  };
  return (
    <TextWithValidationMessages>
      <BasicTextInput
        type={InputType.TEXT}
        onChange={handleOnChange}
        value={value}
        className={value.length === 0 ? "" : 0 < errorMessages.length ? "errorStyle" : "successStyle"}
        {...rest}
      />
      <ul>
        {errorMessages.map((errorMesaage, i) => (
          <li key={i}>{errorMesaage}</li>
        ))}
      </ul>
    </TextWithValidationMessages>
  );
};

TextInputWithValidation.defaultProps = {
  type: InputType.TEXT,
  validations: []
};

export default TextInputWithValidation;
