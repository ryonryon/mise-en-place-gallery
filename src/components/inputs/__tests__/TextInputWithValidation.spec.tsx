import * as React from "react";
import * as renderer from "react-test-renderer";

import TextInputWithValidation from "../TextInputWithValidation";

describe("TextInputWithValidation test", () => {
  it("tests TextInputWithValidation renders correctly ", () => {
    const textInputWithValidation = renderer.create(<TextInputWithValidation />).toJSON();
    expect(textInputWithValidation).toMatchSnapshot();
  });
});
