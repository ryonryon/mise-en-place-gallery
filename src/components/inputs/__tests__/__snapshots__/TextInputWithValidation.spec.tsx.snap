// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputWithValidation test tests TextInputWithValidation renders correctly  1`] = `
<div
  className="css-emrd1e"
>
  <input
    className=" css-1puhmno"
    onChange={[Function]}
    type="text"
    value=""
  />
  <ul />
</div>
`;
